File: Data Structures & Algorithms/the-maze-iii/submission-2.py
class Solution:
    def findShortestWay(self, maze: List[List[int]], ball: List[int], hole: List[int]) -> str:
        # Time Complexity: O((m * n)log(m * n)) Space Complexity: O(m * n)
        minHeap = [] # [[num steps, row, col, path]]
        visited = set()
        moves = ((0,1,'r'),(1,0,'d'),(0,-1,'l'),(-1,0,'u'))
        heapq.heappush(minHeap, [0, ball[0], ball[1], ''])

        while minHeap:
            steps, r, c, path = heapq.heappop(minHeap)
            if (r, c) in visited: continue
            visited.add((r, c))
            if [r, c] == hole: return path
            for dr, dc, p in moves:
                nr, nc = r + dr, c + dc
                cur = 1
                hit_hole = False
                while min(nr, nc) >= 0 and nr < len(maze) and nc < len(maze[nr]) and maze[nr][nc] != 1:
                    if [nr, nc] == hole: 
                        hit_hole = True
                        break
                    nr += dr
                    nc += dc
                    cur += 1
                if not hit_hole:
                    nr -= dr
                    nc -= dc
                    cur -= 1
                if (nr, nc) not in visited: heapq.heappush(minHeap, [steps + cur, nr, nc, path + p])
        return 'impossible'

File: Data Structures & Algorithms/permutation-string/submission-2.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) return false;
        array<int, 27> check1{0};
        for (char c : s1) {
            int cur = c - 'a';
            check1[cur]++;
        }
        array<int, 27> check2{0};
        int l{0}, r{0};
        while (r < s2.size()) {
            check2[s2[r] - (int)'a']++;
            if (check1 == check2) return true;

            r++;
            if (r - l == s1.size()) {
                check2[s2[l] - (int)'a']--;
                l++;
            }
        }
        return false;
    }
};
